File: week08/18/prob3.cpp
#include <iostream>
#include <vector>

using namespace std;

class MyCircularDeque {
public:
    vector<int> q;
    int capacity;
    int size;
    int front;
    int rear;

    MyCircularDeque(int k) {
        q = vector<int>(k, 0);  // Initialize the deque with size k
        capacity = k;
        size = 0;
        front = 0;
        rear = k - 1;  // Rear starts from the last position of the vector
    }
    
    bool insertFront(int value) {
        if (isFull()) return false;
        front = (front - 1 + capacity) % capacity;  // Circular wrap-around for front
        q[front] = value;
        size++;
        return true;
    }
    
    bool insertLast(int value) {
        if (isFull()) return false;
        rear = (rear + 1) % capacity;  // Circular wrap-around for rear
        q[rear] = value;
        size++;
        return true;
    }
    
    bool deleteFront() {
        if (isEmpty()) return false;
        front = (front + 1) % capacity;  // Move front forward, circular wrap-around
        size--;
        return true;
    }
    
    bool deleteLast() {
        if (isEmpty()) return false;
        rear = (rear - 1 + capacity) % capacity;  // Move rear backward, circular wrap-around
        size--;
        return true;
    }
    
    int getFront() {
        if (isEmpty()) return -1;
        return q[front];
    }
    
    int getRear() {
        if (isEmpty()) return -1;
        return q[rear];
    }
    
    bool isEmpty() {
        return size == 0;
    }
    
    bool isFull() {
        return size == capacity;
    }
};

File: week08/26/prob3.py
class MyCircularDeque:

    def __init__(self, k: int):
        self.deque = []
        self.max_len = k
        
    def insertFront(self, value: int) -> bool:
        if len(self.deque) + 1 <= self.max_len:
            self.deque = [value] + self.deque
            return True
        return False

    def insertLast(self, value: int) -> bool:
        if len(self.deque) + 1 <= self.max_len:
            self.deque =  self.deque + [value]
            return True
        return False

    def deleteFront(self) -> bool:
        try:
            self.deque.pop(0)
            return True
        except:
            return False

        

    def deleteLast(self) -> bool:
        try:
            self.deque.pop()
            return True
        except:
            return False
        

    def getFront(self) -> int:
        try:
            return self.deque[0]
        except:
            return -1
        

    def getRear(self) -> int:
        try:
            return self.deque[-1]
        except:
            return -1
        

    def isEmpty(self) -> bool:
        return len(self.deque) == 0
        

    def isFull(self) -> bool:
        return len(self.deque) == self.max_len
        


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()